### PR TITLE
[go_router] Refactors imperative APIs and browser history

### PR DESCRIPTION
**Pull request by chunhtai**
Several thing.

1. I move all the imperative logic from RouterDelegate to RouteInformationParser, so that the imperative API can go through Router parsing pipeline. The Parser will handle modifying mutating RouteMatchList and produce the final RouteMatchList. The RouterDelegate would only focus on building the widget base on the final RouteMatchList 
2. combine RouteMatcher and Redirector with RouteConfiguration. I feel that instead of passing three class instances around, we should probably just have one class for all the route parsing related utility.
3. serialize routeMatchList and store into browser history. This way we can let backward and forward button to reflect imperative operation as well.
4. Some minor clean ups

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by chunhtai** [comment]
ci seems to be weird this time, will try pushing a new commit later

File: packages/go_router/lib/src/parser.dart [hunk 1]
```diff
@@ -128,16 +92,71 @@ class GoRouteInformationParser extends RouteInformationParser<RouteMatchList> {
...
+
+  RouteMatchList _updateRouteMatchList(
+    RouteMatchList newMatchList, {
+    required RouteMatchList? base,
+    required Completer<Object?>? completer,
+    required NavigatingType type,
+  }) {
+    switch (type) {
+      case NavigatingType.push:
+        return base!.push(
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router/lib/src/parser.dart:128]: If they are null, it means something wrong with the GoRouteInformationProvider that provides malformed RouteInformation

File: packages/go_router/lib/src/match.dart [hunk 2]
```diff
@@ -149,3 +146,281 @@ class ImperativeRouteMatch<T> extends RouteMatch {
...
+  }
+
+  @override
+  String toString() {
+    return '${objectRuntimeType(this, 'RouteMatchList')}($fullPath)';
+  }
+}
+
+/// Handles encoding and decoding of [RouteMatchList] objects to a format
+/// suitable for using with [StandardMessageCodec].
```
Comment (johnpryan) [inline packages/go_router/lib/src/match.dart:345]: Is there any benefit / reason this doesn't extend the [Codec](https://api.dart.dev/stable/3.0.3/dart-convert/Codec-class.html) class from dart:convert?

Comment (chunhtai) [reply packages/go_router/lib/src/match.dart:351]: fixed

**Comment by chunhtai** [comment]
Hi @johnpryan, I updated this pr, the change is in the last commit

1. Makes RouteMatchListCodec to implement Codec
2. Move `error` and `extra` property from each RouteMatch to RouteMatchList, since they will be the same for every RouteMatch anyway.

**Review by johnpryan — APPROVED** [review]

**Comment by chunhtai** [comment]
waiting for migration guide to merge https://github.com/flutter/website/pull/8853

**Pull request merged**